

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>predictPAM.main &mdash; predictPAM 1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> predictPAM
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../readme.html">predictPAM: A python module to predict custom PAM sites in any small genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">predictPAM</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">predictPAM</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>predictPAM.main</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for predictPAM.main</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">predictPAM: A python module to predict custom PAM sites in any small genome</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="c1"># Author: Ravin Poudel</span>

<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">string</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">nmslib</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">pyfaidx</span> <span class="k">import</span> <span class="n">Fasta</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="k">import</span> <span class="n">IUPAC</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="k">import</span> <span class="n">permutations</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">pybedtools</span> <span class="k">import</span> <span class="n">BedTool</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">Counter</span>

<div class="viewcode-block" id="myparser"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main.myparser">[docs]</a><span class="k">def</span> <span class="nf">myparser</span><span class="p">():</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;predictPAM: A python module to predict custom PAM sites in any small genome&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--gbkfile&#39;</span><span class="p">,</span> <span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">help</span><span class="o">=</span><span class="s1">&#39;A genbank .gbk file&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--pamseq&#39;</span><span class="p">,</span> <span class="s1">&#39;-p&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;A short PAM motif to search for, may be use IUPAC ambiguous alphabet&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--targetlength&#39;</span><span class="p">,</span> <span class="s1">&#39;-l&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">22</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Length of the target sequence&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--strand&#39;</span><span class="p">,</span> <span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;forward&#39;</span><span class="p">,</span><span class="s1">&#39;reverse&#39;</span><span class="p">],</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;forward&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Strand of DNA&#39;</span><span class="p">)</span> <span class="c1"># use choices array,  use &#39;plus&#39; and &#39;minus&quot;</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--lcp&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Length of conserved sequence close to PAM&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--eds&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">6</span><span class="p">),</span><span class="n">default</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Unexcepted Levenshtein edit distance on the distal portion of target sequence from PAM&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--outfile&#39;</span><span class="p">,</span> <span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;The table of pam sites and data&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--tempdir&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;The temp file directory&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--keeptemp&#39;</span><span class="p">,</span><span class="n">help</span><span class="o">=</span><span class="s2">&quot;Should intermediate files be kept?&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store_true&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--log&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Log file&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;predictPAM.log&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--threads&#39;</span> <span class="p">,</span><span class="n">help</span><span class="o">=</span><span class="s2">&quot;Number of processor threads to use.&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">parser</span></div>

<span class="c1"># GLOBAL parsing variables</span>

<div class="viewcode-block" id="_logger_setup"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main._logger_setup">[docs]</a><span class="k">def</span> <span class="nf">_logger_setup</span><span class="p">(</span><span class="n">logfile</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Set up logging to a logfile and the terminal standard out.</span>

<span class="sd">    Args:</span>
<span class="sd">        logfile (str): Log file</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span><span class="p">,</span>
                            <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> </span><span class="si">%(name)-12s</span><span class="s1"> </span><span class="si">%(levelname)-8s</span><span class="s1"> </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">,</span>
                            <span class="n">datefmt</span><span class="o">=</span><span class="s1">&#39;%m-</span><span class="si">%d</span><span class="s1"> %H:%M&#39;</span><span class="p">,</span>
                            <span class="n">filename</span><span class="o">=</span><span class="n">logfile</span><span class="p">,</span>
                            <span class="n">filemode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
        <span class="c1"># define a Handler which writes INFO messages or higher to the sys.stderr</span>
        <span class="n">console</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">()</span>
        <span class="n">console</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
        <span class="c1"># set a format which is simpler for console use</span>
        <span class="n">formatter</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">Formatter</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1">: </span><span class="si">%(levelname)-8s</span><span class="s1"> </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="c1"># tell the handler to use this format</span>
        <span class="n">console</span><span class="o">.</span><span class="n">setFormatter</span><span class="p">(</span><span class="n">formatter</span><span class="p">)</span>
        <span class="c1"># add the handler to the root logger</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">addHandler</span><span class="p">(</span><span class="n">console</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;An error occurred setting up logging&quot;</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">e</span></div>

<div class="viewcode-block" id="get_fastas"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main.get_fastas">[docs]</a><span class="k">def</span> <span class="nf">get_fastas</span><span class="p">(</span><span class="n">filelist</span><span class="p">,</span> <span class="n">tempdir</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns Fasta and complement of Fasta for a given Genbank file</span>

<span class="sd">    Args:</span>
<span class="sd">        filelist (str): Genbank file to process</span>

<span class="sd">    Returns:</span>
<span class="sd">        forward.fasta(str): Fasta file in forward orientation (5&#39;-3&#39;)</span>
<span class="sd">        reverse.fasta(str): Complement of Fasta file (3&#39;-5&#39;)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">f1</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tempdir</span><span class="p">,</span><span class="s2">&quot;forward.fasta&quot;</span><span class="p">),</span><span class="s2">&quot;w&quot;</span><span class="p">)</span>
        <span class="n">f2</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tempdir</span><span class="p">,</span><span class="s2">&quot;reverse.fasta&quot;</span><span class="p">),</span><span class="s2">&quot;w&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">filelist</span><span class="p">:</span>
            <span class="n">recs</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">recs</span><span class="p">:</span>
                <span class="n">record_f</span> <span class="o">=</span> <span class="n">rec</span>
                <span class="n">record_r</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">complement</span><span class="p">(),</span><span class="n">rec</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="n">rec</span><span class="o">.</span><span class="n">name</span><span class="o">+</span><span class="s2">&quot;_complement&quot;</span><span class="p">,</span><span class="n">rec</span><span class="o">.</span><span class="n">description</span><span class="o">+</span><span class="s2">&quot;_complement&quot;</span><span class="p">)</span>
                <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">record_f</span><span class="p">,</span><span class="n">f1</span><span class="p">,</span><span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
                <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">record_r</span><span class="p">,</span><span class="n">f2</span><span class="p">,</span><span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;An error occurred in input genbank file&quot;</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">e</span></div>


<div class="viewcode-block" id="map_pam"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main.map_pam">[docs]</a><span class="k">def</span> <span class="nf">map_pam</span><span class="p">(</span><span class="n">tempdir</span><span class="p">,</span> <span class="n">pamseq</span><span class="p">,</span> <span class="n">threads</span><span class="p">,</span> <span class="n">strand</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Runs seqkit locate to find the PAM site in the genome (Fasta)</span>

<span class="sd">    Args:</span>
<span class="sd">        pamseq (str) : PAM motif</span>
<span class="sd">        threads (int): Number of processor threads to use</span>
<span class="sd">        strand (str) : forward for 5&#39;-3&#39; or reverse for 3&#39;-5&#39; orientation</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Data in Bedfile format with matches</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">infasta</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tempdir</span><span class="p">,</span> <span class="s2">&quot;forward.fasta&quot;</span><span class="p">)</span>
        <span class="n">infasta_complement</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tempdir</span><span class="p">,</span> <span class="s2">&quot;reverse.fasta&quot;</span><span class="p">)</span>
        <span class="c1"># -i == ignore cases, -d == pattern/motif contains degenerate base</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqkit&quot;</span><span class="p">,</span>
                       <span class="s2">&quot;locate&quot;</span><span class="p">,</span>
                       <span class="s2">&quot;-p&quot;</span><span class="p">,</span> <span class="n">pamseq</span><span class="p">,</span>
                       <span class="n">infasta</span><span class="p">,</span><span class="s2">&quot;-P&quot;</span><span class="p">,</span>
                       <span class="s2">&quot;--bed&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;--threads&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">threads</span><span class="p">),</span>
                        <span class="s2">&quot;-d&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;-i&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;reverse&quot;</span><span class="p">:</span>
            <span class="n">parameters</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">=</span><span class="n">infasta_complement</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">parameters</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span>
        <span class="n">out</span><span class="p">,</span> <span class="n">err</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span>
    <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">e</span>
    <span class="k">except</span> <span class="ne">FileNotFoundError</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">f</span></div>


<div class="viewcode-block" id="get_target"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main.get_target">[docs]</a><span class="k">def</span> <span class="nf">get_target</span><span class="p">(</span><span class="n">tempdir</span><span class="p">,</span> <span class="n">mappingdata</span><span class="p">,</span> <span class="n">targetlength</span><span class="p">,</span> <span class="n">strand</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Given bedfile of PAM locations, function goes up or down and finds target</span>

<span class="sd">    Args:</span>
<span class="sd">        tempdir (str): Temporary directory</span>
<span class="sd">        mappingdata (str): Bedfile output from map_pam</span>
<span class="sd">        targetlength (int) : Length of the target sequence</span>
<span class="sd">        strand (str) : forward for 5&#39;-3&#39; or reverse for 3&#39;-5&#39; orientation</span>
<span class="sd">        </span>
<span class="sd">    Returns:</span>
<span class="sd">        str: Bedfile format with matching genes or CDS</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">target_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># track keys so that any duplicated entry can be removed from the final dictionary</span>
    <span class="n">keys_list</span> <span class="o">=</span><span class="p">[]</span>
    <span class="c1"># this won&#39;t work for big genomes because it reads into memory try seqio index</span>
    <span class="n">infasta</span> <span class="o">=</span> <span class="n">Fasta</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tempdir</span><span class="p">,</span> <span class="s2">&quot;forward.fasta&quot;</span><span class="p">))</span>
    <span class="n">infasta_complement</span> <span class="o">=</span> <span class="n">Fasta</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tempdir</span><span class="p">,</span> <span class="s2">&quot;reverse.fasta&quot;</span><span class="p">))</span>
    <span class="n">bylines</span> <span class="o">=</span> <span class="n">mappingdata</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">bylines</span><span class="p">:</span>
        <span class="n">tline</span> <span class="o">=</span> <span class="n">entry</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="n">whichchromose</span><span class="o">=</span><span class="n">tline</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">pam_sp</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tline</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span> <span class="c1"># -1 to adjust- because seqkit convention - starts from 1 but in python starts from 0.</span>
        <span class="n">pam_ep</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tline</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="n">pam_seq</span> <span class="o">=</span> <span class="n">tline</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
        <span class="n">seqid</span> <span class="o">=</span> <span class="n">infasta</span><span class="p">[</span><span class="n">whichchromose</span><span class="p">]</span><span class="o">.</span><span class="n">name</span>
        <span class="c1"># note the strand is not mean + from seqkit mapping. Comes from user- orientation of genome to search for target</span>
        <span class="k">if</span> <span class="n">strand</span><span class="o">==</span><span class="s2">&quot;forward&quot;</span><span class="p">:</span>
            <span class="n">target_sp</span> <span class="o">=</span> <span class="n">pam_sp</span> <span class="o">-</span> <span class="n">targetlength</span>
            <span class="n">target_ep</span> <span class="o">=</span> <span class="n">pam_sp</span>
            <span class="n">target_seq</span> <span class="o">=</span> <span class="n">infasta</span><span class="p">[</span><span class="n">whichchromose</span><span class="p">][</span><span class="n">target_sp</span><span class="p">:</span><span class="n">target_ep</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
        <span class="k">if</span> <span class="n">strand</span><span class="o">==</span><span class="s2">&quot;reverse&quot;</span><span class="p">:</span>
            <span class="n">target_sp</span> <span class="o">=</span> <span class="n">pam_ep</span>
            <span class="n">target_ep</span> <span class="o">=</span> <span class="n">pam_ep</span> <span class="o">+</span> <span class="n">targetlength</span>
            <span class="n">target_seq</span> <span class="o">=</span> <span class="n">infasta_complement</span><span class="p">[</span><span class="n">whichchromose</span><span class="p">][</span><span class="n">target_sp</span><span class="p">:</span><span class="n">target_ep</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">target_seq</span><span class="p">)</span> <span class="o">==</span> <span class="n">targetlength</span><span class="p">:</span>
                <span class="n">target_dict</span><span class="p">[</span><span class="n">target_seq</span><span class="p">]</span><span class="o">=</span> <span class="p">{</span><span class="s2">&quot;seqid&quot;</span><span class="p">:</span> <span class="n">seqid</span><span class="p">,</span> <span class="s2">&quot;target_sp&quot;</span><span class="p">:</span> <span class="n">target_sp</span><span class="p">,</span>
                <span class="s2">&quot;target_ep&quot;</span><span class="p">:</span> <span class="n">target_ep</span><span class="p">,</span> <span class="s2">&quot;pam_seq&quot;</span><span class="p">:</span> <span class="n">pam_seq</span><span class="p">,</span>
                 <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">}</span>
        <span class="n">keys_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">target_seq</span><span class="p">)</span>
    <span class="n">remove_target</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">Counter</span><span class="p">(</span><span class="n">keys_list</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">]</span> <span class="c1"># list of keys with more than one observation</span>
    <span class="n">target_dict2</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">target_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">k</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">remove_target</span><span class="p">}</span> <span class="c1"># although dict over writes on non-unique key, but we want to complete remove such observation</span>
    <span class="k">return</span> <span class="n">target_dict2</span></div>


<span class="c1"># parse key then create first section and remaining part, also take care of strand specificity</span>
<div class="viewcode-block" id="parse_target"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main.parse_target">[docs]</a><span class="k">def</span> <span class="nf">parse_target</span><span class="p">(</span><span class="n">targetdict</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">seqlengthtopam</span><span class="p">):</span> <span class="c1"># use local variable for close12</span>
    <span class="sd">&quot;&quot;&quot;Given a dictionary of target sequence, parse target sequence into two parts:</span>
<span class="sd">    close region and remaining seq, then create a new dictionary with unique close region sequences as keys</span>

<span class="sd">    Args:</span>
<span class="sd">        targetdict (dict): Dictionary of target sequences obtained after running get_target</span>
<span class="sd">        strand (str) : forward for 5&#39;-3&#39; or reverse for 3&#39;-5&#39; orientation</span>
<span class="sd">        seqlengthtopam (int) : Length of conserved sequence close to PAM</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict : Dictionary with unique close region sequences as keys</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parse_target_dict</span><span class="o">=</span><span class="p">{}</span>
    <span class="c1"># track keys so that any duplicated entry can be removed from the final dictionary</span>
    <span class="n">keys_list</span> <span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">items</span> <span class="ow">in</span> <span class="n">targetdict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">items</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;forward&quot;</span><span class="p">:</span>
            <span class="n">proxitopam</span> <span class="o">=</span> <span class="n">items</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="o">-</span><span class="n">seqlengthtopam</span><span class="p">:]</span>
            <span class="n">distaltopam</span> <span class="o">=</span> <span class="n">items</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="o">-</span><span class="n">seqlengthtopam</span><span class="p">]</span>
            <span class="n">items</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;target&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">items</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># move target sequence (key) as value</span>
            <span class="n">items</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;distaltopam&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">distaltopam</span>
            <span class="n">items</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;proxitopam&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">proxitopam</span>
            <span class="n">parse_target_dict</span><span class="p">[</span><span class="n">proxitopam</span><span class="p">]</span> <span class="o">=</span> <span class="n">items</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="c1">## will retain only target with unique bp, make proxitopam as key</span>
            <span class="n">keys_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">proxitopam</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;reverse&quot;</span><span class="p">:</span>
            <span class="n">proxitopam</span> <span class="o">=</span> <span class="n">items</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="n">seqlengthtopam</span><span class="p">]</span>
            <span class="n">distaltopam</span> <span class="o">=</span> <span class="n">items</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">seqlengthtopam</span><span class="p">:]</span>
            <span class="n">items</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;target&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">items</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">items</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;distaltopam&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">distaltopam</span>
            <span class="n">items</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;proxitopam&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">proxitopam</span>
            <span class="n">parse_target_dict</span><span class="p">[</span><span class="n">proxitopam</span><span class="p">]</span> <span class="o">=</span> <span class="n">items</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">keys_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">proxitopam</span><span class="p">)</span>
    <span class="n">remove_target</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">Counter</span><span class="p">(</span><span class="n">keys_list</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">parse_target_dict2</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">parse_target_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">k</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">remove_target</span><span class="p">}</span>
    <span class="k">return</span> <span class="n">parse_target_dict2</span></div>

<span class="c1">#nms lib - create a index with remainingseq</span>
<div class="viewcode-block" id="create_index"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main.create_index">[docs]</a><span class="k">def</span> <span class="nf">create_index</span><span class="p">(</span><span class="n">strings</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Initializes and returns a NMSLIB index</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        strings (str): Strings to calculate distance</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">        index (nmslib.dist.IntIndex): Returns a NMSLIB index</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">index</span> <span class="o">=</span> <span class="n">nmslib</span><span class="o">.</span><span class="n">init</span><span class="p">(</span><span class="n">space</span><span class="o">=</span><span class="s1">&#39;leven&#39;</span><span class="p">,</span>
                        <span class="n">dtype</span><span class="o">=</span><span class="n">nmslib</span><span class="o">.</span><span class="n">DistType</span><span class="o">.</span><span class="n">INT</span><span class="p">,</span>
                        <span class="n">data_type</span><span class="o">=</span><span class="n">nmslib</span><span class="o">.</span><span class="n">DataType</span><span class="o">.</span><span class="n">OBJECT_AS_STRING</span><span class="p">,</span>
                        <span class="n">method</span><span class="o">=</span><span class="s1">&#39;small_world_rand&#39;</span><span class="p">)</span>
    <span class="n">index</span><span class="o">.</span><span class="n">addDataPointBatch</span><span class="p">(</span><span class="n">strings</span><span class="p">)</span>
    <span class="n">index</span><span class="o">.</span><span class="n">createIndex</span><span class="p">(</span><span class="n">print_progress</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">index</span></div>

<div class="viewcode-block" id="filter_parse_target"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main.filter_parse_target">[docs]</a><span class="k">def</span> <span class="nf">filter_parse_target</span><span class="p">(</span><span class="n">parse_dict</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">levendistance</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a list of features for a genbank file</span>

<span class="sd">    Args:</span>
<span class="sd">        parse_dict(dict): A dictionary with parse target sequence</span>
<span class="sd">        threads (int) : Number of processor threads to use</span>
<span class="sd">        levendistance (int) : Unexcepted Levenshtein edit distance on the distal portion of target sequence from PAM</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of features with genebank informations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filter_parse_dict</span><span class="o">=</span><span class="p">{}</span>
    <span class="c1"># get a list of ramaning sequences</span>
    <span class="n">rms_list</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">key_val</span> <span class="ow">in</span> <span class="n">parse_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">rms_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_dict</span><span class="p">[</span><span class="n">key_val</span><span class="p">][</span><span class="s1">&#39;distaltopam&#39;</span><span class="p">])</span>
    <span class="c1"># initialize a new index</span>
    <span class="n">ref_index</span> <span class="o">=</span> <span class="n">create_index</span><span class="p">(</span><span class="n">rms_list</span><span class="p">)</span>
    <span class="c1"># Find knn of 1 for each remaning seq</span>
    <span class="k">for</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">parse_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">query_string</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="s1">&#39;distaltopam&#39;</span><span class="p">]</span>
        <span class="n">ids</span><span class="p">,</span> <span class="n">distances</span> <span class="o">=</span> <span class="n">ref_index</span><span class="o">.</span><span class="n">knnQuery</span><span class="p">(</span><span class="n">query_string</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> <span class="c1">## k =number of k nearest neighbours (knn)</span>
        <span class="k">if</span> <span class="n">distances</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">levendistance</span><span class="p">:</span> <span class="c1"># check the second value, because the first value will be mostly zero, distance between the same query and index</span>
            <span class="n">filter_parse_dict</span><span class="p">[</span><span class="n">keys</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
    <span class="k">return</span> <span class="n">filter_parse_dict</span></div>


<div class="viewcode-block" id="get_genbank_features"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main.get_genbank_features">[docs]</a><span class="k">def</span> <span class="nf">get_genbank_features</span><span class="p">(</span><span class="n">filelist</span><span class="p">):</span>
    <span class="c1">## Pseudocode</span>
    <span class="sd">&quot;&quot;&quot;Return a list of features for a genbank file</span>

<span class="sd">    Args:</span>
<span class="sd">        filelist(genebank): Genbank file to process</span>


<span class="sd">    Returns:</span>
<span class="sd">        list: List of features with genebank informations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">feature_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">filelist</span><span class="p">:</span>
        <span class="n">genebank_file</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">file</span><span class="p">,</span><span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">genebank_file</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">entry</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                <span class="n">feature_dict</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;CDS&#39;</span><span class="p">,</span> <span class="s1">&#39;gene&#39;</span><span class="p">]:</span>
                    <span class="n">feature_dict</span><span class="p">[</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">entry</span><span class="o">.</span><span class="n">id</span>
                    <span class="n">feature_dict</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">position</span>
                    <span class="n">feature_dict</span><span class="p">[</span><span class="s2">&quot;stop&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span><span class="o">.</span><span class="n">position</span>
                    <span class="n">feature_dict</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">type</span>
                    <span class="n">feature_dict</span><span class="p">[</span><span class="s2">&quot;strand_for_feature&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;reverse&#39;</span> <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">strand</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;forward&#39;</span>
                    <span class="k">for</span> <span class="n">qualifier_key</span><span class="p">,</span> <span class="n">qualifier_val</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                        <span class="n">feature_dict</span><span class="p">[</span><span class="n">qualifier_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">qualifier_val</span>
                    <span class="n">feature_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">feature_dict</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">feature_list</span></div>


<span class="c1"># ######### pybedtools ########</span>

<div class="viewcode-block" id="get_nearby_feature"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main.get_nearby_feature">[docs]</a><span class="k">def</span> <span class="nf">get_nearby_feature</span><span class="p">(</span><span class="n">target_mappingfile</span><span class="p">,</span> <span class="n">featurefile</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Adds downstream information to the given target sequences and mapping information</span>

<span class="sd">    Args:</span>
<span class="sd">        target_mappingfile (str) : Data in Bedfile format with matches</span>
<span class="sd">        featurefile (list) : List of features with genebank informations</span>

<span class="sd">    Returns:</span>
<span class="sd">        (tsv?? check type of pybed output): A file with target sequences, mapping information, and downstream information</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># format to featurefile</span>
    <span class="n">featurebed</span> <span class="o">=</span> <span class="n">BedTool</span><span class="p">(</span><span class="n">featurefile</span><span class="o">.</span><span class="n">splitlines</span><span class="p">())</span>
    <span class="c1"># format to mapping file</span>
    <span class="n">mapbed</span> <span class="o">=</span> <span class="n">BedTool</span><span class="p">(</span><span class="n">target_mappingfile</span><span class="o">.</span><span class="n">splitlines</span><span class="p">())</span>
    <span class="c1"># get feature downstream of target sequence</span>
    <span class="n">downstream</span> <span class="o">=</span> <span class="n">mapbed</span><span class="o">.</span><span class="n">closest</span><span class="p">(</span><span class="n">featurebed</span> <span class="p">,</span> <span class="n">d</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fd</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">D</span><span class="o">=</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)</span>
    <span class="c1"># get feature upstream of target sequence</span>
    <span class="n">upstream</span> <span class="o">=</span> <span class="n">mapbed</span><span class="o">.</span><span class="n">closest</span><span class="p">(</span><span class="n">featurebed</span> <span class="p">,</span> <span class="n">d</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">D</span><span class="o">=</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">downstream</span><span class="p">,</span> <span class="n">upstream</span></div>


<div class="viewcode-block" id="merge_downstream_upstream"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main.merge_downstream_upstream">[docs]</a><span class="k">def</span> <span class="nf">merge_downstream_upstream</span><span class="p">(</span><span class="n">downsfile</span><span class="p">,</span><span class="n">upsfile</span><span class="p">,</span><span class="n">columns_name</span><span class="p">,</span> <span class="n">outputfilename</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a merged file</span>

<span class="sd">    Args:</span>
<span class="sd">        (tsv?? check type of pybed output): A file with target sequences, mapping information, and upstream information</span>
<span class="sd">        (tsv?? check type of pybed output): A file with target sequences, mapping information, and downstream information</span>

<span class="sd">    Returns:</span>
<span class="sd">        dataframe: A DataFrame with merged upstream information and downstream information for a target sequence</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">downstream_df</span> <span class="o">=</span> <span class="n">downsfile</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">(</span><span class="n">names</span><span class="o">=</span><span class="n">columns_name</span><span class="p">,</span><span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">upstream_df</span> <span class="o">=</span> <span class="n">upsfile</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">(</span><span class="n">names</span><span class="o">=</span><span class="n">columns_name</span><span class="p">,</span><span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">all_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">downstream_df</span><span class="p">,</span> <span class="n">upstream_df</span><span class="p">,</span>
                      <span class="n">right_on</span><span class="o">=</span><span class="n">columns_name</span><span class="p">[:</span><span class="mi">8</span><span class="p">],</span>
                      <span class="n">left_on</span><span class="o">=</span><span class="n">columns_name</span><span class="p">[:</span><span class="mi">8</span><span class="p">],</span>
                      <span class="n">how</span><span class="o">=</span><span class="s1">&#39;outer&#39;</span><span class="p">)</span>
    <span class="n">all_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span><span class="n">outputfilename</span><span class="p">),</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">all_df</span></div>
<span class="c1">################</span>
<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../predictPAM.html#predictPAM.main.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Run Complete predictPAM workflow.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Set up logging</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">myparser</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="p">:</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="n">_logger_setup</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">log</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">pamseq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">pamseq</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">unambiguous_dna</span><span class="p">)</span>
        <span class="c1">#parse genbank file: todo allow multiple Genbank files</span>
        <span class="c1">#record = SeqIO.parse(args.gbkfile, &quot;genbank&quot;)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">tempdir</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">tempdir</span><span class="p">):</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Specified location for tempfile (</span><span class="si">{}</span><span class="s2">) does not exist, using default location.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tempdir</span><span class="p">))</span>
                <span class="n">tempdir</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">mkdtemp</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;pamPredict_&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">tempdir</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">mkdtemp</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;pamPredict_&#39;</span><span class="p">,</span> <span class="nb">dir</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">tempdir</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Temp directory is: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">tempdir</span><span class="p">)</span>


        <span class="c1"># Try to avoid writing out and reading genome in fasta format</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Retriving fastas- forward and reverse from a genbanke file&quot;</span><span class="p">)</span>
        <span class="n">get_fastas</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">gbkfile</span><span class="p">,</span> <span class="n">tempdir</span><span class="o">=</span><span class="n">tempdir</span><span class="p">)</span>

        <span class="c1"># mapping</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Mapping pam to the genome&quot;</span><span class="p">)</span>
        <span class="n">mapfile</span> <span class="o">=</span> <span class="n">map_pam</span><span class="p">(</span><span class="n">tempdir</span><span class="o">=</span><span class="n">tempdir</span><span class="p">,</span> <span class="n">pamseq</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">pamseq</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">threads</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>

        <span class="c1"># Retrieving target sequence</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Retrieving target sequence for matching PAM : </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">tempdir</span><span class="p">)</span>
        <span class="n">targetdict</span> <span class="o">=</span> <span class="n">get_target</span><span class="p">(</span><span class="n">tempdir</span><span class="o">=</span><span class="n">tempdir</span><span class="p">,</span> <span class="n">mappingdata</span><span class="o">=</span><span class="n">mapfile</span><span class="p">,</span> <span class="n">targetlength</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">targetlength</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>

        <span class="c1"># Parsing target sequence into two: 1)close12 and 2) remainingseq</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parsing target sequence into two: 1)proxitopam and 2) distaltopam&quot;</span><span class="p">)</span>
        <span class="n">parsetargetdict</span> <span class="o">=</span> <span class="n">parse_target</span><span class="p">(</span><span class="n">targetdict</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span> <span class="n">seqlengthtopam</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">lcp</span><span class="p">)</span>

        <span class="c1"># Calculate Levenshtein distance among remainingseq, and remove any remainingseq that are similar</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Filtering parse target sequence based on Levenshtein distance using NMSLIB index, with knn=1&quot;</span><span class="p">)</span>
        <span class="n">filterparsetargetdict</span> <span class="o">=</span> <span class="n">filter_parse_target</span><span class="p">(</span><span class="n">parsetargetdict</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">threads</span><span class="p">,</span> <span class="n">levendistance</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">eds</span><span class="p">)</span>

        <span class="c1"># reformating filterparsetargetdict- to make compatible for pybed</span>
        <span class="n">filterparsetargetdict_pd</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">filterparsetargetdict</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">)</span>
        <span class="c1"># remove index, which is key of dict</span>
        <span class="n">filterparsetargetdict_pd_dindx</span> <span class="o">=</span> <span class="n">filterparsetargetdict_pd</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># pybed takes tab separated file with no header, plus first three column has to be as above</span>
        <span class="n">filterparsetargetdict_pd_dindx_tab</span> <span class="o">=</span> <span class="n">filterparsetargetdict_pd_dindx</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># get get_genbank_features from a genebank file</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Retrieving CDS/gene information for each record in the genebank file&quot;</span><span class="p">)</span>
        <span class="n">genebankfeatures</span> <span class="o">=</span> <span class="n">get_genbank_features</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">gbkfile</span><span class="p">)</span>
        <span class="c1"># reformating genebankfeatures- to make compatible for pybed</span>
        <span class="n">genebankfeatures_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">genebankfeatures</span><span class="p">)</span>
        <span class="c1"># enpty tab crates isses in runnign pybed, so replance NaN with NA, then make tab separated</span>
        <span class="n">genebankfeatures_df</span> <span class="o">=</span> <span class="n">genebankfeatures_df</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s2">&quot;NA&quot;</span><span class="p">)</span>
        <span class="n">genebankfeatures_df_tab</span> <span class="o">=</span> <span class="n">genebankfeatures_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># pybedtools</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Mapping features downstream and upstream of target sequence&quot;</span><span class="p">)</span>
        <span class="n">down</span><span class="p">,</span> <span class="n">up</span> <span class="o">=</span> <span class="n">get_nearby_feature</span><span class="p">(</span><span class="n">filterparsetargetdict_pd_dindx_tab</span><span class="p">,</span> <span class="n">genebankfeatures_df_tab</span><span class="p">)</span>

        <span class="c1">### Adding column name to upstream and downstream output from pybed.</span>
        <span class="c1"># get columns from two input file: target_mappingfile(filterparsetargetdict), featurefile(genebankfeatures_df)</span>
        <span class="n">targetfile_columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">filterparsetargetdict</span><span class="o">.</span><span class="n">values</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">featurefile_columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">genebankfeatures_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="c1"># from genebank feature dataframe</span>
        <span class="n">joined_columns</span> <span class="o">=</span> <span class="n">targetfile_columns</span> <span class="o">+</span> <span class="n">featurefile_columns</span>
        <span class="n">joined_columns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;distance&quot;</span><span class="p">)</span> <span class="c1"># on top of mapping upstream and downstream, we are recording distance in pybed closest. Thus need an extra column</span>

        <span class="c1"># merge upstream and downstream output</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Merging downstream and upstream features.Columns with suffix _x represents information from downstream whereas suffix with _y represent that of upstream&quot;</span><span class="p">)</span>
        <span class="n">merged_down_ups</span> <span class="o">=</span> <span class="n">merge_downstream_upstream</span><span class="p">(</span><span class="n">down</span><span class="p">,</span><span class="n">up</span><span class="p">,</span><span class="n">joined_columns</span><span class="p">,</span><span class="n">outputfilename</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">outfile</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;predictPAM terminated with errors. See the log file for details.&quot;</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="k">raise</span> <span class="ne">SystemExit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">finally</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">keeptemp</span><span class="p">:</span>
                <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="n">tempdir</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">UnboundLocalError</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="k">pass</span></div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Ravin Poudel, Adam R. Rivers

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>